site/_posts/2020-11-13-securing-flutter-oauth.md

```diff
@@ -0,0 +1,766 @@
+
+After successfully getting the access token, the `getUserDetails` function is called to get user information and set the required state variables, which changes the UI rendered.
+
+The library makes things pretty easy. But you might have a couple of questions. Why not use a form to capture the user credentials? Well, we want to delegate all auth operations to FusionAuth. This means we don't have to worry as much about securing our native app, because credential management happens within the external user management server.
+
+OK, that's fine. But if you run this code, you'll see it pops out to the system browser. Why not use an embedded webview? Well, because a webview is totally controlled by the native application displaying it, the [current mobile best practices](https://tools.ietf.org/html/rfc8252) for OAuth require you to use the system browser, which is not under that control. From section 8.12 of that document:
+
+> This best current practice requires that native apps MUST NOT use embedded user-agents to perform authorization requests and allows that authorization endpoints MAY take steps to detect and block authorization requests in embedded user-agents.  
+
+After that digression, here's `loginAction`:
+
+```dart
+Future<void> loginAction() async {
+  setState(() {
+    isBusy = true;
+    errorMessage = '';
+  });
+
+  try {
+    final AuthorizationTokenResponse result =
+        await appAuth.authorizeAndExchangeCode(
+      AuthorizationTokenRequest(
+        FUSION_AUTH_CLIENT_ID,
+        FUSION_AUTH_REDIRECT_URI,
+        issuer: 'https://$FUSION_AUTH_DOMAIN',
+        scopes: <String>['openid'],
+      ),
+    );
+    log('data: $result');
+    final Map<String, Object> profile =
+        await getUserDetails(result.accessToken);
+
+    debugPrint('response: $profile');
+    await secureStorage.write(
+          key: 'refresh_token', value: result.refreshToken);
+    var gravatar = Gravatar(profile['email']);
+    var url = gravatar.imageUrl(
+      size: 100,
+      defaultImage: GravatarImage.retro,
+      rating: GravatarRating.pg,
+      fileExtension: true,
+    );
+    setState(() {
+      isBusy = false;
+      isLoggedIn = true;
+      name = profile['given_name'];
+      picture = url;
+    });
+  } on Exception catch (e, s) {
+    debugPrint('login error: $e - stack: $s');
+
+    setState(() {
+      isBusy = false;
+      isLoggedIn = false;
+      errorMessage = e.toString();
+    });
+  }
+}
+// ...
+```
+
+`loginAction` called `getUserDetails`. What does this to? It takes the access token as a parameter and makes the HTTP GET request to the auth server. If successful, it will receive the user details in response. This is a standard OIDC call. This function is called inside the `initAction` function as well.
```

Comment:
I think the first sentence is a bit short. Should it have some type of intro like, "As you can see in the code,"